corgidrp/sorting.py

```diff
@@ -0,0 +1,232 @@
+        for subset in split_exptime[0]:
+            for frame in subset:
+                frame_id_list += [extract_frame_id(frame.filename)]
+                exptime_list += [frame.ext_hdr['EXPTIME']]
+                unity_gain_filepath_list += [frame.filepath]
+        idx_id_sort = np.argsort(frame_id_list)
+        exptime_arr = np.array(exptime_list)[idx_id_sort]
+        # Count repeated, consecutive elements
+        count_cons = [1]
+        exptime_cons = [exptime_arr[0]]
+        idx_cons = 0
+        for exptime in exptime_arr:
+            if exptime == exptime_cons[-1]:
+                count_cons[idx_cons] += 1
+            else:
+                idx_cons += 1
+                count_cons += [1]
+                exptime_cons += [exptime]
+        # First index always has a repetition in the previous loop (id=id)
+        count_cons[0] -= 1
+
+        idx_cons2 = [0]
+        exptime_cons2 = [exptime_arr[idx_cons2[0]]]
+        kgain_subset = []
+        # Iterate over unique counts
+        for idx_count in range(len(count_cons) - 1):
+            if count_cons[idx_count+1] == count_cons[idx_count]:
+                exptime_cons2 += [exptime_cons[idx_count+1]]
+                idx_cons2 += [idx_count+1]
+            else:
+                # Last exposure time must be repeated and only once
+                if (exptime_cons2[-1] in exptime_cons2[:-1] and
+                    len(exptime_cons2) - 1 == len(set(exptime_cons2))):
+                    kgain_subset += [idx_cons2[1] - 1, idx_cons2[-1]]
+                    idx_cons2 = [idx_count+1]
+                    exptime_cons2 = [exptime_arr[idx_cons2[0]]]
+                else:
+                # It is not a subset for kgain
+                    continue
+        # Choose the largest subset
+        kgain_subset = np.array(kgain_subset)
+        idx_kgain = np.argmax(kgain_subset[1::2] - kgain_subset[0::2])
+        # Extract first/last index in the subset of consecutive frames
+        idx_kgain_0 = kgain_subset[2 * idx_kgain]
+        idx_kgain_1 = kgain_subset[2 * idx_kgain + 1]
+        # Count frames before and subset length
+        idx_kgain_first = np.sum(count_cons[0:idx_kgain_0])
+        idx_kgain_last = idx_kgain_first + np.sum(count_cons[idx_kgain_0:idx_kgain_1+1])
+
+        # Sort unity gain filenames
+        unity_gain_filepath_arr = np.array(unity_gain_filepath_list)[idx_id_sort]
+        cal_list = unity_gain_filepath_arr[idx_kgain_first:idx_kgain_last]
+        # Update OBSTYPE and take profit to check files are in the list
+        n_kgain = 0
+        cal_frame_list = []
+        for frame in dataset_cp:
+            if frame.filepath in cal_list:
+                frame.pri_hdr['OBSTYPE'] = 'KGAIN'
+                cal_frame_list += [frame]
+                n_kgain += 1
+
+        print(f'K-gain has {n_kgain} unity frames with exposure times', exptime_cons[idx_kgain_0:idx_kgain_1+1], f'seconds with each {count_cons[idx_kgain_0]} frames each')
```

Comment:
this print statement and the nonlin one below seem like good things to write into the header

Reply:
I added it to the HISTORY.

Example when run for K-gain:
Dataset to calibrate K-GAIN. A sorting algorithm based on the constraints that NFRAMES, EXPTIME and CMDGAIN have when collecting calibration data for K-gain, Non-linearity and EM-gain vs DAC was applied to an input dataset from PUPILIMG visit files.The result is that K-gain has 110 unity frames with exposure times [0.077, 0.77, 1.538, 2.308, 3.077, 3.846, 4.615, 5.385, 6.154, 6.923, 7.692, 8.462, 9.231, 10.0, 10.769, 11.538, 12.308, 13.077, 13.846, 14.615, 15.385, 1.538] seconds with 5 frames each.

Example when run for non-linearity:
Dataset to calibrate NON-LIN. A sorting algorithm based on the constraints that NFRAMES, EXPTIME and CMDGAIN have when collecting calibration data for K-gain, Non-linearity and EM-gain vs DAC was applied to an input dataset from PUPILIMG visit files.The result is that Non-linearity has 242 frames with gains [1.65, 5.24, 8.6, 16.7, 27.5, 45.26, 87.5, 144.1, 237.26, 458.7, 584.4].

Reply:
Perfect. This is exactly what we are looking to put into the history.